### PR TITLE
Handle the no-target case through user information.

### PR DESCRIPTION
**Pull request by olivier-stasse**
Goal
====

Handling package with no targets (typically data packages)
Issue #398

Remplacement of PR #403

Principle
=======

The user specifies the property/variable ZERO_TARGET of the target PROJECT_NAME  to true:
```cmake
SET(${PROJECT_NAME}_ZERO_TARGET TRUE)
```

Then the export of exportTarget.cmake is not generated (currently generating a bug),
and then an expression is set on Config.cmake.in avoiding the inclusion of this file.
This is necessary to avoid later error when upstream packages are using the cmake recipe of this package.

When a normal package is used the expression is true and thus the inclusion of the target is done.

Tested in both situations.

### CODE REVIEW

File: Config.cmake.in [hunk 1]
```diff
@@ -108,7 +108,10 @@ IF(COMMAND ADD_REQUIRED_DEPENDENCY)
   ENDFOREACH()
 ENDIF(COMMAND ADD_REQUIRED_DEPENDENCY)
 
-include("${CMAKE_CURRENT_LIST_DIR}/@TARGETS_EXPORT_NAME@.cmake")
+@EXPR_PROJECT_NAME_ZERO_TARGET@
+if(NOT @PROJECT_NAME@_ZERO_TARGET)
+  include("${CMAKE_CURRENT_LIST_DIR}/@TARGETS_EXPORT_NAME@.cmake")
+endif()
```
Comment (jmirabel) [inline Config.cmake.in:114]: What about the simpler version
```suggestion
@INCLUDE_TARGETS_FILE@
```

File: package-config.cmake [hunk 2]
```diff
@@ -154,6 +154,13 @@ if(_PKG_CONFIG_REQUIRES)
   endforeach()
   list(REMOVE_DUPLICATES _PKG_CONFIG_REQUIRES_LIST)
 endif(_PKG_CONFIG_REQUIRES)
+
+if(NOT ${PROJECT_NAME}_ZERO_TARGET)
+  SET(EXPR_PROJECT_NAME_ZERO_TARGET "SET(${PROJECT_NAME}_ZERO_TARGET FALSE)")  
+else()
+  SET(EXPR_PROJECT_NAME_ZERO_TARGET "SET(${PROJECT_NAME}_ZERO_TARGET TRUE)")
+endif()
```
Comment (jmirabel) [inline package-config.cmake:162]:
```suggestion
if(NOT ${PROJECT_NAME}_ZERO_TARGET)
  SET(INCLUDE_TARGETS_FILE "include(\"\${CMAKE_CURRENT_LIST_DIR}/${TARGETS_EXPORT_NAME}.cmake\")")  
else()
  SET(INCLUDE_TARGETS_FILE "# Package with no targets")
endif()
```

**Comment by olivier-stasse** [comment]
Nice simplification. I just tested it, it worked as proposed.

**Comment by jmirabel** [comment]
I let @gergondet merge if it is ok for him now.

**Comment by gergondet** [comment]
This looks fine, but:

1. The variable should be documented in `base.cmake`
2. It's a little strange to prefix the variable name with `${PROJECT_NAME}` where other such variables use the `PROJECT_` prefix (see https://jrl-cmakemodules.readthedocs.io/en/master/pages/base.html#required-variables)
3. I think `PROJECT_EXPORT_NO_TARGET` is a better name
4. `@INCLUDE_TARGETS_FILE@` could be replaced with:
```cmake
if(EXISTS "${CMAKE_CURRENT_LIST_DIR}/@TARGETS_EXPORT_NAME@.cmake") 
  include("${CMAKE_CURRENT_LIST_DIR}/@TARGETS_EXPORT_NAME@.cmake")
endif()
```

**Comment by jmirabel** [comment]
>     4\. `@INCLUDE_TARGETS_FILE@` could be replaced with:
> 
> 
> ```cmake
> if(EXISTS "${CMAKE_CURRENT_LIST_DIR}/@TARGETS_EXPORT_NAME@.cmake") 
>   include("${CMAKE_CURRENT_LIST_DIR}/@TARGETS_EXPORT_NAME@.cmake")
> endif()
> ```

I thought about that too. I preferred the explicit method so that, at configure time, you explicitly say whether or not there is a targets file, rather than looking for it later. This avoids the annoying issue of a file existing but not found. If there is one but you did not find it, it is better to report an error.

**Comment by olivier-stasse** [comment]
Last commit is correcting point 1, 2 and 3.
I let you decide on point 4.
But my inclination would be in the direction of Joseph.

File: package-config.cmake [hunk 3]
```diff
@@ -154,6 +154,13 @@ if(_PKG_CONFIG_REQUIRES)
   endforeach()
   list(REMOVE_DUPLICATES _PKG_CONFIG_REQUIRES_LIST)
 endif(_PKG_CONFIG_REQUIRES)
+
+if(NOT PROJECT_EXPORT_NO_TARGET)
+  SET(INCLUDE_TARGETS_FILE "include(\"\${CMAKE_CURRENT_LIST_DIR}/${TARGETS_EXPORT_NAME}.cmake\")")
```
Comment (gergondet) [inline package-config.cmake:159]: -set+SET (same below)

**Comment by gergondet** [comment]
I agree on point 4 as well. So just two more small changes and it's good too go :)

**Comment by olivier-stasse** [comment]
SET -> set done

**Pull request merged**